OpenOversight/app/commands.py

```diff
@@ -411,6 +421,388 @@ def bulk_add_officers(filename, no_create, update_by_name, update_static_fields)
+                if counter % 1000 == 0:
+                    print("Processed {} incidents.".format(counter))
+            print("Done with incidents. Processed {} rows.".format(counter))
+
+    counter = 0
+    if links_csv is not None:
+        with open(links_csv) as f:
+            csv_reader = csv.DictReader(f)
+            csv_reader.fieldnames = _unify_field_names(csv_reader.fieldnames)
+            _check_provided_fields(
+                csv_reader,
+                required_fields=["id", "url"],
+                optional_fields=[
+                    "title",
+                    "link_type",
+                    "description",
+                    "author",
+                    "user_id",
+                    "officer_ids",
+                    "incident_ids",
+                ],
+                csv_name="links",
+            )
+            existing_officer_links = (
+                Link.query.join(Link.officers)
+                .filter(Officer.department_id == department_id)
+                .all()
+            )
+            existing_incident_links = (
+                Link.query.join(Link.incidents)
+                .filter(Incident.department_id == department_id)
+                .all()
+            )
+            id_to_link = {
+                link.id: link
+                for link in existing_officer_links + existing_incident_links
+            }
+            for row in csv_reader:
+                row["officers"] = _objects_from_split_field(
+                    row.get("officer_ids"), all_officers
+                )
+                row["incidents"] = _objects_from_split_field(
+                    row.get("incident_ids"), all_incidents
+                )
+                _create_or_update_model(
+                    row=row,
+                    existing_model_lookup=id_to_link,
+                    create_method=create_link_from_dict,
+                    update_method=update_link_from_dict,
+                    force_create=force_create,
+                    model=Link,
+                )
+                counter += 1
+                if counter % 1000 == 0:
+                    print("Processed {} links.".format(counter))
+            print("Done with links. Processed {} rows.".format(counter))
+
+    db.session.commit()
+    print("All committed.")
+
+    if force_create:
+        # This will only work in postgres and fail in sqlite
+        raw_sql = """
+        select setval('officers_id_seq', (select max(id) from officers));
+        select setval('salaries_id_seq', (select max(id) from salaries));
+        select setval('assignments_id_seq', (select max(id) from assignments));
+        select setval('links_id_seq', (select max(id) from links));
+        select setval('incidents_id_seq', (select max(id) from incidents));
+        """
+        try:
+            db.session.execute(raw_sql)
+            print("Updated sequences.")
+        except SQLAlchemyError:
+            print("Failed to update sequences")
+
```

Comment:
this seems like a lot of custom ETL code in general, especially to bake into a CLI too. It mostly looks good, but could some more-etl specific framework help us here, e.g. pandas or something? compared to vanilla python? Also should this script be in the CLI tool specifically? Will it remain tenable?

Reply:
Definitely a good question. I can't think of a way of doing his with pandas, but I am certainly open to suggestions. Specifically I don't know of solutions (within pandas or any ETL tool) to import several tables and maintain foreign key relationships between them (since the respective primary keys are only decided once loaded into the database)

Reply:
Also, I do agree that this code shouldn't all live in the `commands.py`. I will see where I can move it